controller_manager/src/resource_manager.cpp

```diff
@@ -0,0 +1,143 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+#include <vector>
+#include <utility>
+
+#include "hardware_interface/components/component_info.hpp"
+#include "hardware_interface/component_parser.hpp"
+#include "hardware_interface/hardware_info.hpp"
+#include "hardware_interface/actuator_hardware.hpp"
+#include "hardware_interface/sensor_hardware.hpp"
+#include "hardware_interface/system_hardware.hpp"
+
+#include "pluginlib/class_loader.hpp"
+
+#include "./resource_manager.hpp"
+
+namespace controller_manager
+{
+
+class ResourceStorage
+{
+  static constexpr const char * pkg_name = "hardware_interface";
+  static constexpr const char * actuator_interface_name =
+    "hardware_interface::ActuatorHardwareInterface";
+  static constexpr const char * sensor_interface_name =
+    "hardware_interface::SensorHardwareInterface";
+  static constexpr const char * system_interface_name =
+    "hardware_interface::SystemHardwareInterface";
+
+public:
+  ResourceStorage()
+  : actuator_loader_(pkg_name, actuator_interface_name),
+    sensor_loader_(pkg_name, sensor_interface_name),
+    system_loader_(pkg_name, system_interface_name)
+  {}
+
+  ~ResourceStorage() = default;
+
+  template<class HardwareT, class HardwareInterfaceT>
+  void initialize_hardware(
+    const hardware_interface::HardwareInfo & hardware_info,
+    pluginlib::ClassLoader<HardwareInterfaceT> & loader,
+    std::vector<HardwareT> & container)
+  {
+    // hardware_class_type has to match class name in plugin xml description
+    auto interface = std::unique_ptr<HardwareInterfaceT>(
+      loader.createUnmanagedInstance(hardware_info.hardware_class_type));
+    HardwareT actuator(std::move(interface));
+    container.emplace_back(std::move(actuator));
+    container.back().configure(hardware_info);
+  }
+
+  void initialize_actuator(const hardware_interface::HardwareInfo & hardware_info)
+  {
+    using HardwareT = hardware_interface::ActuatorHardware;
```

Comment:
I am no sure this `using` are really needed. Since there is only one use of the "defines" it doesn't increase readability.

Reply:
it's not really needed, but I actually introduced it exactly for the reason of increasing readability ;-)

Reply:
I had the same thought, thinking there are some other options that look slightly better
```
    using hardware_interface::ActuatorHardware;
    using hardware_interface::ActuatorHardwareInterface;
    initialize_hardware<ActuatorHardware, ActuatorHardwareInterface>(
      hardware_info, actuator_loader_, actuators_);
```
the same a little bolder 

```
    using namespace hardware_interface;
    initialize_hardware<ActuatorHardware, ActuatorHardwareInterface>(
      hardware_info, actuator_loader_, actuators_);
```
or as a slightly dodgy "one-liner"
```
    initialize_hardware<
      hardware_interface::ActuatorHardware, 
      hardware_interface::ActuatorHardwareInterface>(
        hardware_info, actuator_loader_, actuators_);
```

Reply:
In terms of readability I personally found the more explicit forms better, in this case the "business logic" is your single function call. While it may be appealing to create a repeating pattern, it's better to make things more explicit locally, if someone needs to follow a trace of calls in this codebase, or only interested in this section, they shouldn't be introduced to the beauty in other parts of this file.